### ISSUE TITLE
[Feature Request] Audio Left/Right Channel Slider on Audio Sources

### ISSUE THREAD

**Issue by Enverex** [Frooxius/NeosPublic#421]
Currently you can control the volume of audio sources, change their doppler level, etc but there is no way to change the left/right channel bias. e.g. to be able to change it so that you can only hear the left audio channel of the sound source.

Example of actual use: I was setting up speakers in a room and wanted to pipe the audio out of the speakers themselves rather than the video "display". This can be done, but both speakers play both left and right channels which sounds odd when you're stood near them and wouldn't be expected in a real-world scenario. Being able to set the left speaker to only play the left channel and the right speaker to only play the right channel would fix that.

**Comment by shiftyscales** [comment]
I could also see this as being potentially useful for multi-channel audio support in the future. E.g. being able to set up a more fully-featured/simulated theater.

**Comment by Frooxius** [comment]
Hmm, yeah there's a bit of a question on how to best achieve this. I think this might be better done as a filter that goes between the audio source and the audio output, because that gives more flexibility (e.g. you can have multi-channel filter or use things like downmixing or using the single channel for whole.

If you want the audio to come out of the actual speakers, then you're likely to use them with spatialization, but that process itself already converts the audio to mono first before spatializing.

**Comment by Enverex** [comment]
Yeah, in that example I'm already spacialising, but it's then piping both channels to one point in mono and it creates a weird (and strangely uncomfortable) effect. I wouldn't recommend it.

Being able to select channels, as Shifty mentioned, would then allow you to even create proper theatres with 5.1 - 9.1 surround systems using speakers in the virtual space., which would also be a nice addition.

**Comment by Ruikio** [comment]
Now that audio is being worked on, now would be a good time to revive this topic. Having proper theaters inside Neos would be amazing. Also could do some amazing things with immersion inside maps.

**Comment by shiftyscales** [comment]
For another sound-related query, would you consider prioritizing the implementation of multi-channel filtering as per this issue, or other types of aforementioned/previously requested audio filters, e.g. ring modulator/pitch shifting filters, @Frooxius?

**Comment by Frooxius** [comment]
I'm sorry, but this actually needs an audio filtering / DSP system first, so it can be built on top of that. That would allow chaining audio sources and putting filters in between the source and the actual output (e.g. one that extracts a single channel), but Neos is currently missing framework for this.

It could be prioritized potentially, but that's pretty big system, so it would take a while to put in place before that can get added.

**Comment by shiftyscales** [comment]
Ah, I see. I had found a quote in Discord where I had saw you say:
>Pitch shifting is a bit more complex audio processing, so that will require some audio filters to be implemented.
>
>The audio system in Neos is designed to allow for insertion of filters between the source of audio data and the audio output, but I haven't added any filters yet.
>
>But yeah making it high or low pitched or generally messing with your voice is going to be lots of fun!  I'd want to add ring modulator at some point too and make a Dalek avatar.

So I had figured it was just a matter of implementing the filters themselves. Thanks for clarifying, @Frooxius!

There could be some momentum/excitement raised from providing additional tooling/development in our audio capabilities- but probably not at the expense of other upcoming systems on the roadmap.

**Comment by clarjon1** [comment]
I didn't spot this when opening 3474 earlier, so I'll add in my two cents here:

A UI for placing spatialized channels, perhaps little speaker icons that get spawned when clicking an additional button relating to an audio source to spatial channel sources for placement, and be able to assign them, either thru logix or inspector UI, to the appropriate audio channel (eg left/right/center/rear-left/rear-right/etc)

Users would then be able to place the spatial channel sources around their scenes as they desire, allowing for effects and audio to come from multiple areas in the scene, all from a single audio file or video file or streaming video source, and interact with them with logix as expected

This would allow for video stream players to be able to look at the incoming audio and set the available channels, outputting to the preplaced options as an example, and corresponding logix might be able to see that a video only supports left/right/center audio and set the tagged sources as expected, a mono source could set all tagged to use the same source for the appropriate effect, or whatever handling is coded in by the tool creator

**Comment by troyBORG** [comment]
I guess since my thing got merged I to this.
Mines more in regard to the component "LocalAudioDeviceStream" not just audio in general.

Like right now it seems to be be choosing an audio source.  Id like be be able to maybe say
Audio from voice (since there is the whole main, multimedia, music, voice, sfx) channels.
Then be able to pick left or right.  Plug that I into a volume meter to get it as a local value for myself for driving gradient drivers for my ears perking up directionly to people talking.

Currently if music is playing, or someone sends me a invite that sfx will make them move.
